### PR TITLE
Multiple owls

### PR DESCRIPTION
**Pull request by invisime**
This changeset enables multiple owl games as follows:

- default initialization of 6 owls
- optional initialization of a specific number of owls
- moves require an indication of which owl to move
- the Player classes pick which owl to move
- moves respect the "hoot" rule (jumping instead of sharing a space)

### CODE REVIEW

File: GameEngine/GameState.cs [hunk 1]
```diff
@@ -38,10 +38,10 @@ public void TakeTurn()
             }
             else
             {
-                var cardToPlay = Player.SelectCardToPlay(Board);
-                Console.WriteLine("Player chose to play a {0} card", cardToPlay);
-                Board.Move(cardToPlay);
-                Player.Discard(cardToPlay);
+                var play = Player.FormulatePlay(Board);
+                Console.WriteLine("Player chose to play a {0} card on an owl at position {1}", play.Card, play.Position);
```
Comment (rpherbig) [inline GameEngine/GameState.cs]: 💡 It feels like `Play` should have its own `ToString` that is used here

File: GameEngine/GameBoard.cs [hunk 2]
```diff
@@ -57,19 +53,39 @@ private BoardPositionType Convert(CardType cardType)
...
+                throw new InvalidMoveException("There is no owl at position " + position);
+            }
+            return owlIndex;
+        }
+
+        private int FindDestinationPosition(Play play)
+        {
+            var desiredBoardColor = Convert(play.Card);
+            int newPosition = play.Position;
+            Predicate<int> isEmptySpaceOfCorrectColor = (position) =>
```
Comment (rpherbig) [inline GameEngine/GameBoard.cs]: 💡 What do you think about extracting this into an instance method?

File: GameEngine/GameBoard.cs [hunk 3]
```diff
@@ -1,45 +1,41 @@
...
-                if(Board[i] == targetBoardPosition || Board[i] == BoardPositionType.Nest)
-                {
-                    OwlPosition = i;
-                    break;
-                }
-            }
+            var owlToMove = FindOwl(play.Position);
+            var newPosition = FindDestinationPosition(play);
+
+            OwlPositions[owlToMove] = newPosition;
```
Comment (rpherbig) [inline GameEngine/GameBoard.cs]: 💡 I'm liking the idea of that `Parliament` class more now that I see this indexing

File: GameEngine/GameBoard.cs [hunk 4]
```diff
@@ -1,45 +1,41 @@
...
 namespace GameEngine
 {
     public class GameBoard
     { 
         public List<BoardPositionType> Board { get; private set; }
-        public int OwlPosition { get; set; }
-
-
-        public GameBoard(int gameSizeMultiplier)
+        public List<int> OwlPositions { get; set; }
```
Comment (rpherbig) [inline GameEngine/GameBoard.cs]: 💡  Can the setter be `private`?

File: GameEngineTests/GameBoardTests.cs [hunk 5]
```diff
@@ -25,41 +25,112 @@ public void ShouldStartOwlsAtFirstSpots()
...
+            var newOwlPosition = board.OwlPositions[0];
+            Assert.AreEqual(board.NestPosition, newOwlPosition,
+                "Owl is not on last position.");
+            Assert.AreEqual(BoardPositionType.Nest, board.Board[newOwlPosition],
+                "Owl's position is not a nest.");
         }
 
         [TestMethod]
-        public void ShouldFailToMoveOwlWhenItIsAlreadyInTheNest()
+        public void ShouldFailToMoveWhenThereAreNoOwlsAtThePosition()
```
Comment (rpherbig) [inline GameEngineTests/GameBoardTests.cs:68]: ❓ The intent behind `ShouldFailToMoveOwlWhenItIsAlreadyInTheNest` was that you cannot make a `Play` which targets an Owl already in the Nest. I don't think that is captured in this test.

Could you add a new test for this case? A quick glance at the code makes me think the game will currently allow this move which it should not.

File: GameEngine/GameBoard.cs [hunk 6]
```diff
@@ -1,45 +1,41 @@
...
+        public int NestPosition
+        {
+            get { return Board.Count - 1; }
+        }
+        
+        public GameBoard(int gameSizeMultiplier, int numberOfOwls = 6)
         {
             BuildBoard(gameSizeMultiplier);
-            OwlPosition = 0;
+            OwlPositions = Enumerable.Repeat(0, numberOfOwls).ToList();
```
Comment (rpherbig) [inline GameEngine/GameBoard.cs]: ❌ The Owls do not all start on the same space. They start in positions 0 to N.

File: GameEngine/GameBoard.cs (hunk 2 above)
Comment (rpherbig) [reply GameEngine/GameBoard.cs]: Or maybe keeping it here and adding the `newPosition < NestPosition` check to the `Predicate`...?

File: GameEngine/Players/Player.cs [hunk 7]
```diff
@@ -21,11 +22,24 @@ public void Discard(CardType card)
...
-        public abstract CardType SelectCardToPlay(GameBoard board);
-
         public bool HandContainsSun()
         {
             return Hand.Contains(CardType.Sun);
         }
+
+        public abstract Play FormulatePlay(GameBoard board);
+
+        protected int PositionOfStragglerOwl(GameBoard board)
```
Comment (rpherbig) [inline GameEngine/Players/Player.cs]: 💡 These two don't really feel to me like they belong here. It feels a bit premature to guess what other `Players` will want in terms of helpers.

**Comment by rpherbig** [comment]
Sorry, I'm a little rusty with GitHub and made a bunch of single comments instead of one review.

💡 Idea, discussion, doesn't hold up merge.
❓ Question needs answered before merge.
❌ Issue needs fixed before merge.

The changes look great. One ❓ and one ❌ (and I'll own up to the fact the bug was in my original commit...).

File: GameEngine/GameState.cs (hunk 1 above)
Comment (invisime) [reply GameEngine/GameState.cs]: Easy enough. Done.

File: GameEngine/Players/Player.cs (hunk 7 above)
Comment (invisime) [reply GameEngine/Players/Player.cs]: These have been moved to the `Parliament` class.

File: GameEngine/GameBoard.cs (hunk 3 above)
Comment (invisime) [reply GameEngine/GameBoard.cs]: Added it. It uses a `HashSet` under the hood. It's working great!

File: GameEngine/GameBoard.cs (hunk 6 above)
Comment (invisime) [reply GameEngine/GameBoard.cs]: I have updated the tests and implementation to reflect this.

File: GameEngineTests/GameBoardTests.cs (hunk 5 above)
Comment (invisime) [reply GameEngineTests/GameBoardTests.cs:68]: Restored a version of `ShouldFailToMoveOwlWhenItIsAlreadyInTheNest`.

File: GameEngine/GameBoard.cs (hunk 4 above)
Comment (invisime) [reply GameEngine/GameBoard.cs]: It can, and is now.

File: GameEngine/GameBoard.cs (hunk 2 above)
Comment (invisime) [reply GameEngine/GameBoard.cs]: `FindDestinationPosition` has been refactored. The new version incorporates these suggestions.

**Pull request merged**